### PR TITLE
Support partition by

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue  #4749

## What this PR does / why we need it:
Record the partition information of the created table into the metadata table, and display the partition information of the table through the show statement

### CODE REVIEW

**Review by daviszhen — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Review by fengttt — COMMENTED** [review]
So the partition info in output in comment?   Why can't we output as part of DDL in proper syntax?

**Comment by nnsgmsone** [comment]
Why parition just a string?

**Comment by qingxinhome** [comment]
> Why parition just a string?

At present, we support partition:
1. Support the syntax of MySQL partition clause and
2. Support semantic check of MySQL partition clauses
3. Record partition clause information and partition expression into metadata
4. The data storage layer does not support partition temporarily, so the partition information is expressed in string

**Review by iamlinjunhong — APPROVED** [review]

File: pkg/sql/plan/partition.go [hunk 1]
```diff
@@ -73,6 +73,7 @@ func buildHashPartition(partitionBinder *PartitionBinder, partitionOp *tree.Part
 		return err
 	}
 
+	partitionInfo.PartitionMsg = tree.String(partitionOp, dialect.MYSQL)
```
Comment (ouyuanning) [inline pkg/sql/plan/partition.go:76]: use tree.String is not a good idea。
some special string like ',  "   will removed in tree.String
you can not get origin PartitionOp from this string.

File: proto/plan.proto [hunk 2]
```diff
@@ -246,6 +246,7 @@ message PartitionInfo {
 	repeated PartitionItem partitions = 7;
 	int64 algorithm = 8;
 	bool isSubPartition = 9;
+	string partitionMsg = 10;
```
Comment (ouyuanning) [inline proto/plan.proto:249]: suggest use "_" to split words in  proto file

**Review by XuPeng-SH — APPROVED** [review]
lgtm

**Review by aressu1985 — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

File: proto/plan.proto (hunk 2 above)
Comment (qingxinhome) [reply proto/plan.proto]: ok.no problem!

File: pkg/sql/plan/partition.go (hunk 1 above)
Comment (qingxinhome) [reply pkg/sql/plan/partition.go:76]: There is no good processing strategy for this, and the string () method of the abstract syntax tree can be reconstructed later

**Review by ouyuanning — APPROVED** [review]

**Pull request merged**